In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [36]:
#Importing the required dataset
df=pd.read_csv("loan-dataset-1.csv")

Data Exploration

In [5]:
df.head()

loan_id  principal_loan_amount request_date  \
0  70280c1b-4339-4949-ba69-ad2a6cb6b782                   8880   2023-03-06   
1  8a550559-9749-491b-b6aa-7188bc0a488f                  13720   2023-05-10   
2  5644d27e-e05e-4a40-bef3-64187892338f                   2090   2023-02-16   
3  9d8d17ac-8af9-4936-ba5d-8d5ed5e0cfab                   6820   2023-02-26   
4  4586ad7a-2231-419b-b8e3-ba03b929b56e                  13770   2023-04-13   

   interest_rate loan_channel is_approved is_default  customer_monthly_income  \
0            9.0          WEB        True        NaN                     6100   
1           10.0   MOBILE_APP        True        NaN                     5150   
2            7.5        AGENT        True        NaN                     6030   
3            8.5   MOBILE_APP         NaN        NaN                     2150   
4           10.0    AFFILIATE         NaN        NaN                     5810   

   customer_age customer_gender  customer_credit_score  customer_zip_code  
0            30            Male                      3              10007  
1            39          Female                      3              10004  
2            43          Female                      3              10003  
3            54            Male                      0              10004  
4            28          Female                      2              10007

In [6]:
df.tail()

loan_id  principal_loan_amount  \
15158  1a3c6a20-4da0-422c-8c89-d9d1906b2f84                   3320   
15159  c870947f-28f3-424e-89d4-ebc0d8126bcb                  12370   
15160  fde0b44b-f82b-4452-8719-8190e95a5884                  14150   
15161  91bd0c9b-306c-41ea-b6b1-aa7b01c71b5c                   3250   
15162  0608c5bc-2eaa-42c2-a034-3d189fda6567                  10990   

      request_date  interest_rate    loan_channel is_approved is_default  \
15158   2023-05-03            8.0           AGENT        True        NaN   
15159   2023-01-03            9.5  DIRECT_SELLING        True        NaN   
15160   2023-05-19           10.0       AFFILIATE        True        NaN   
15161   2023-01-22            8.0  DIRECT_SELLING        True        NaN   
15162   2023-01-21            9.5       AFFILIATE        True        NaN   

       customer_monthly_income  customer_age customer_gender  \
15158                     3900            49            Male   
15159                     3120            33               M   
15160                     3120            31          Female   
15161                     6390            38            Male   
15162                     3090            43            Male   

       customer_credit_score  customer_zip_code  
15158                      5              10005  
15159                      3              10009  
15160                      4              10007  
15161                      3              10009  
15162                      3              10005

In [37]:
#No. of rown and No. of columns
df.shape

(15163, 12)

In [9]:
#Total no. of cells
df.size

181956

In [12]:
#Name of columns in dataset
df.columns

Index(['loan_id', 'principal_loan_amount', 'request_date', 'interest_rate',
       'loan_channel', 'is_approved', 'is_default', 'customer_monthly_income',
       'customer_age', 'customer_gender', 'customer_credit_score',
       'customer_zip_code'],
      dtype='object')

In [14]:
#Data_types of each column
df.dtypes

loan_id                     object
principal_loan_amount        int64
request_date                object
interest_rate              float64
loan_channel                object
is_approved                 object
is_default                  object
customer_monthly_income      int64
customer_age                 int64
customer_gender             object
customer_credit_score        int64
customer_zip_code            int64
dtype: object

In [15]:
#for count of non_null values in a column (helps identifying which column has missing values)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15163 entries, 0 to 15162
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_id                  15163 non-null  object 
 1   principal_loan_amount    15163 non-null  int64  
 2   request_date             15163 non-null  object 
 3   interest_rate            15163 non-null  float64
 4   loan_channel             15163 non-null  object 
 5   is_approved              11771 non-null  object 
 6   is_default               545 non-null    object 
 7   customer_monthly_income  15163 non-null  int64  
 8   customer_age             15163 non-null  int64  
 9   customer_gender          15163 non-null  object 
 10  customer_credit_score    15163 non-null  int64  
 11  customer_zip_code        15163 non-null  int64  
dtypes: float64(1), int64(5), object(6)
memory usage: 1.4+ MB


In [79]:
#no. of rows associated with a loan_id (helps identifying multiple applicants)
df['loan_id'].value_counts()

9dc23377-2007-4bd4-8e70-aba5d4ca6b77    2
8440c973-b1d6-4cf7-98aa-053ee5de4b7a    2
047f4180-ce70-4739-9951-5396b914d243    2
d8613add-4898-4513-a583-a17b02e4e25d    2
c8c480d2-a01f-4c64-8dc9-e5851e3d03b4    2
                                       ..
f752d063-ad75-4e20-91db-160beea7d593    1
582e8286-a428-4181-ba00-4e3625f82ed8    1
23113018-36c5-4a53-b1a5-18a89f20caac    1
24c51ce7-910b-4991-a943-e834b6fa639d    1
0608c5bc-2eaa-42c2-a034-3d189fda6567    1
Name: loan_id, Length: 15000, dtype: int64

In [29]:
#To determine Range of column(max and mean)
df.describe()

principal_loan_amount  interest_rate  customer_monthly_income  \
count           15163.000000   15163.000000             15163.000000   
mean             8072.432236       8.852800              4253.158346   
std              4050.341783       0.814432              1304.176559   
min              1000.000000       7.500000              2000.000000   
25%              4560.000000       8.000000              3120.000000   
50%              8110.000000       9.000000              4250.000000   
75%             11590.000000       9.500000              5390.000000   
max             15000.000000      10.000000              6500.000000   

       customer_age  customer_credit_score  customer_zip_code  
count  15163.000000           15163.000000       15163.000000  
mean      41.070896               3.440018       10005.838027  
std       11.200741               1.367703           2.098677  
min       22.000000               0.000000       10001.000000  
25%       31.000000               3.000000       10004.000000  
50%       41.000000               4.000000       10006.000000  
75%       51.000000               5.000000       10008.000000  
max       60.000000               5.000000       10009.000000

In [33]:
#All intrest rates at which loan is being given
df['interest_rate'].unique()

array([ 9. , 10. ,  7.5,  8.5,  9.5,  8. ])

In [48]:
#All the unique principal amounts
df['principal_loan_amount'].unique()

array([ 8880, 13720,  2090, ...,  8560,  1720,  1180])

In [57]:
#All unique dates
df['request_date'].unique()

array(['2023-03-06', '2023-05-10', '2023-02-16', '2023-02-26',
       '2023-04-13', '2023-02-22', '2023-04-21', '2023-05-27',
       '2023-01-10', '2023-03-17', '2023-04-29', '2023-01-13',
       '2023-01-17', '2023-06-24', '2023-04-23', '2023-04-17',
       '2023-01-06', '2023-03-20', '2023-03-09', '2023-02-20',
       '2023-02-05', '2023-04-03', '2023-06-26', '2023-05-31',
       '2023-05-11', '2023-04-08', '2023-06-27', '2023-03-10',
       '2023-01-24', '2023-03-02', '2023-02-07', '2023-02-01',
       '2023-01-08', '2023-05-18', '2023-04-12', '2023-05-05',
       '2023-06-10', '2023-04-14', '2023-05-14', '2023-02-08',
       '2023-06-29', '2023-03-25', '2023-04-15', '2023-04-07',
       '2023-01-22', '2023-05-07', '2023-06-14', '2023-06-15',
       '2023-05-23', '2023-06-02', '2023-02-17', '2023-05-22',
       '2023-05-24', '2023-06-09', '2023-05-28', '2023-06-07',
       '2023-02-24', '2023-01-03', '2023-02-18', '2023-05-13',
       '2023-03-08', '2023-03-16', '2023-01-16', '2023-

In [58]:
#different loan channels in dataset
df['loan_channel'].unique()

array(['WEB', 'MOBILE_APP', 'AGENT', 'AFFILIATE', 'DIRECT_SELLING',
       'website_revamped', 'ANDROID_V2', 'ANDROID_V3', 'apple_new_v3'],
      dtype=object)

In [61]:
df['is_approved'].unique()

array([True, nan], dtype=object)

In [62]:
df['is_default'].unique()

array([nan, True], dtype=object)

In [65]:
df['customer_monthly_income'].unique()

array([6100, 5150, 6030, 2150, 5810, 4780, 3510, 3090, 5520, 2830, 2190,
       5720, 3470, 2790, 5470, 2110, 4330, 5080, 6090, 4240, 5060, 5570,
       4540, 4200, 2380, 2500, 5880, 2780, 5160, 3120, 2290, 4400, 5260,
       5620, 5360, 2870, 3370, 2200, 5940, 2800, 5690, 3100, 3600, 3000,
       4710, 6420, 4380, 4150, 6470, 3830, 3980, 3080, 4440, 3500, 5250,
       4570, 2950, 6360, 5320, 5010, 5490, 2980, 4230, 2060, 4760, 5200,
       5120, 5440, 3680, 3720, 2340, 2050, 6260, 4910, 2520, 5180, 6070,
       3440, 3270, 3860, 4830, 4740, 3030, 4010, 4390, 5040, 4500, 6280,
       3750, 4430, 4270, 4450, 3760, 3220, 6240, 4180, 6060, 5030, 3850,
       2900, 5300, 4000, 2580, 2590, 2260, 5770, 2460, 5680, 2430, 2270,
       3310, 6250, 4990, 2370, 5450, 2170, 5020, 3970, 5290, 4480, 2440,
       2030, 5050, 6320, 4620, 2670, 3530, 3420, 6460, 5560, 3060, 5580,
       2010, 2420, 4610, 3520, 6270, 5370, 5590, 6150, 2490, 2480, 6410,
       5330, 5780, 3620, 6110, 5910, 5340, 2630, 25

In [66]:
#Unique ages
df['customer_age'].unique()

array([30, 39, 43, 54, 28, 42, 49, 35, 44, 47, 22, 56, 53, 45, 46, 59, 50,
       23, 48, 58, 31, 41, 34, 24, 37, 40, 51, 57, 36, 38, 52, 33, 25, 29,
       32, 26, 27, 60, 55])

In [67]:
#Genders in dataset (needs to be fixed )
df['customer_gender'].unique()

array(['Male', 'Female', 'M', 'F'], dtype=object)

In [68]:
#Range of credit score
df['customer_credit_score'].unique()

array([3, 0, 2, 1, 4, 5])

In [69]:
df['customer_zip_code'].unique()

array([10007, 10004, 10003, 10008, 10009, 10002, 10005, 10006, 10001])

In [35]:
#No. of loans given at what interest rates
df['interest_rate'].value_counts()

10.0    2798
9.0     2774
9.5     2689
8.5     2664
8.0     2629
7.5     1609
Name: interest_rate, dtype: int64

In [70]:
#Hown many loan applications on a respective date
df['request_date'].value_counts()

2023-04-15    175
2023-04-16    152
2023-02-10    104
2023-03-04    104
2023-06-23    101
             ... 
2023-06-12     63
2023-06-11     63
2023-06-22     62
2023-01-05     62
2023-04-03     57
Name: request_date, Length: 180, dtype: int64

In [71]:
df['is_approved'].value_counts()

True    11771
Name: is_approved, dtype: int64

In [72]:
df['is_default'].value_counts()

True    545
Name: is_default, dtype: int64

In [73]:
#applicants with same income
df['customer_monthly_income'].value_counts()

5660    51
2800    51
3580    50
5100    49
2820    48
        ..
4040    19
4410    19
4350    19
5380    16
6500     2
Name: customer_monthly_income, Length: 451, dtype: int64

In [74]:
df['customer_age'].value_counts()

45    448
31    428
44    422
59    418
28    410
48    410
54    406
33    403
29    403
50    399
35    399
47    399
27    397
40    396
37    396
34    394
58    393
57    392
32    392
49    388
46    387
52    383
26    382
53    379
41    379
22    379
25    378
55    378
38    377
39    377
60    376
51    374
43    373
42    368
23    365
30    355
56    354
36    353
24    353
Name: customer_age, dtype: int64

In [75]:
#How many males and how many females
df['customer_gender'].value_counts()

Female    7111
Male      7101
M          518
F          433
Name: customer_gender, dtype: int64

In [76]:
#No. of people with each credit score
df['customer_credit_score'].value_counts()

3    4889
5    3935
4    3828
1     841
0     837
2     833
Name: customer_credit_score, dtype: int64

In [77]:
df['customer_zip_code'].value_counts()

10003    2278
10006    2163
10005    2140
10008    2103
10004    2097
10009    2050
10007    1971
10002     190
10001     171
Name: customer_zip_code, dtype: int64

In [59]:
df['loan_channel'].value_counts()

WEB                 3165
AGENT               3046
AFFILIATE           3041
DIRECT_SELLING      2899
MOBILE_APP          2791
website_revamped      84
ANDROID_V3            58
apple_new_v3          40
ANDROID_V2            39
Name: loan_channel, dtype: int64

In [60]:
df['request_date'].value_counts()

2023-04-15    175
2023-04-16    152
2023-02-10    104
2023-03-04    104
2023-06-23    101
             ... 
2023-06-12     63
2023-06-11     63
2023-06-22     62
2023-01-05     62
2023-04-03     57
Name: request_date, Length: 180, dtype: int64

In [64]:
df['is_approved'].value_counts()

True    11771
Name: is_approved, dtype: int64

In [63]:
df['is_default'].value_counts()

True    545
Name: is_default, dtype: int64

In [45]:
#column wise missing values identification
df.isnull().sum()

loan_id                        0
principal_loan_amount          0
request_date                   0
interest_rate                  0
loan_channel                   0
is_approved                 3392
is_default                 14618
customer_monthly_income        0
customer_age                   0
customer_gender                0
customer_credit_score          0
customer_zip_code              0
dtype: int64

In [43]:
#No. of missing values in Column "is_approved"
df['is_approved'].isnull().sum()

3392

In [47]:
#No. of missing values in Column "is_default"
df['is_default'].isnull().sum()

14618

In [87]:
#Are there any duplicate columns in the dataset
df.duplicated().any()


True

In [90]:
if df.duplicated().any():
    print('There are', df.duplicated().sum(), 'duplicate rows in the DataFrame\n')
else:
    print('There are no duplicate rows in the DataFrame\n')

There are 163 duplicate rows in the DataFrame



In [95]:
df.isna().any()

loan_id                    False
principal_loan_amount      False
request_date               False
interest_rate              False
loan_channel               False
is_approved                 True
is_default                  True
customer_monthly_income    False
customer_age               False
customer_gender            False
customer_credit_score      False
customer_zip_code          False
dtype: bool

Total rows :  15163 

There are 163 duplicate rows in the DataFrame

There are NaN values in the DataFrame
is_approved    True
is_default     True
dtype: bool 

principal_loan_amount: 1000 to 15000

request_date: 2023-01-01 to 2023-06-29

interest_rate: 7.5 to 10.0

loan_channel: ['WEB' 'MOBILE_APP' 'AGENT' 'AFFILIATE' 'DIRECT_SELLING'
 'website_revamped' 'ANDROID_V2' 'ANDROID_V3' 'apple_new_v3']
 
is_approved: [True nan]

is_default: [nan True]

customer_monthly_income: 2000 to 6500

customer_age: 22 to 60

customer_gender: ['Male' 'Female' 'M' 'F']

In [97]:
#Data Cleaning

#Dropping duplicate rows
df=df.drop_duplicates()

In [100]:
print("Total rows :", df.shape[0])

Total rows : 15000


In [108]:
if df.duplicated().any():
    print("There are",df.duplicated().sum(),"duplicated rows in data frame")
else:
    print("There are no duplicate rows in data frame")

There are no duplicate rows in data frame


In [119]:
#Changing empty cells in columns to False
df['is_approved'].fillna(False, inplace=True)
df['is_default'].fillna(False, inplace=True)

if df.isnull().any().any():
    print("There are", df.isnull().sum().sum(),"NAN VALUES in data set")
else:
    print("There are no Nan values in data farme")

There are no Nan values in data farme


0